projects/frontend/src/draft-posts/index.md

```diff
@@ -0,0 +1,148 @@
+
+## Capturing Metrics 
+Not too long ago I was asked to build a demo for one of our clients at work. The task was to capture the response metrics of a web service and add those metrics to [Prometheus](https://prometheus.io/) for display in [Grafana](https://grafana.com/). The only catch was that I did not have direct access to the web service and therefore couldn’t instrument the code. I completed the task by adapting the go reverse proxy to observe the network traffic coming from the service.
+
+Let’s extend our simple proxy to read and report metrics about the downstream service responses. To do this we’ll return to the  _httputil.ReverseProxy_ struct once more. It exposes a  struct field _ReverseProxy.ModifyResponse_ which gives us access to the http response before it goes back to the client.
+
+```go
+ModifyResponse func(*net/http.Response) error
+```
+
+HTTP bodies in go are implemented as _io.Reader_’s and therefore they can only be read once. If you would like to parse them before forwarding the request/response you will need to copy the body into a byte buffer and the reset the request/response body. An obvious drawback is that we buffer the entire response in memory without limit. This could lead to memory issues in production if you received a large response but for our use case this wasn’t an issue. Here’s a quick implementation to duplicate the response body.
+
+```go
+func duplicateReader(in io.ReadCloser) (b, b1 io.ReadCloser, err error) {
+	bBytes, err := ioutil.ReadAll(in)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	in.Close()
+
+	buf := bytes.NewBuffer(bBytes)
+	b = ioutil.NopCloser(buf)
+	b1 = ioutil.NopCloser(bytes.NewBuffer(bBytes))
+
+	return
+}
+```
+
+Depending on your use case you may not actually want to duplicate the reader. You may opt for returning the parsed body of the response i.e. parsing the json into a struct or _map[string]interface{}_  and just resetting _req.Body_ to an equivalent reader.
+
+With the request body problem solved, capturing metrics is simple.
+
+```go
+	proxy := httputil.NewSingleHostReverseProxy(u)
+	// Configure proxy.Director ...
+
+	// ModifyResponse runs before forwarding the downstream response back to the client
+	proxy.ModifyResponse = func(res *http.Response) error {
+			var err error
+			var body io.ReadCloser
+
+			body, res.Body, err = duplicateReader(res.Body)
+			if err != nil {
+				panic(err)
+			}
+
+			fmt.Printf("body = %+v\n", body)
+
+			captureMetrics(body)
+			return nil
+		}
+}
+```
+
+And that’s it! The capture metrics function was pretty specific to my use case so I’ll leave it up to you to implement. Form here it is pretty straightforward to further extend for applications like load balancing, redirect policies, etc.
```

Comment:
```suggestion
And that’s it! The capture metrics function was specific to my use case so I’ll leave it up to you to implement. From here it is pretty straightforward to further extend for applications like load balancing, redirect policies, etc.
```

Reply:
Also, I do not believe it is simple to extend for load balancing. You implemented a proxy with a single backend. You should link to a multi-backend standard library version if available.